noteable_magics/sql/meta_commands.py

```diff
@@ -483,6 +501,353 @@ def _split_schema_table(schema_table: str) -> Tuple[Optional[str], str]:
+        """Introspect all indexes for a table or materialized view, describing the results as a List[IndexModel]"""
+        indexes = []
+
+        index_dicts = inspector.get_indexes(relation_name, schema_name)
+
+        for index_dict in sorted(index_dicts, key=lambda d: d['name']):
+
+            indexes.append(
+                IndexModel(
+                    name=index_dict['name'],
+                    columns=index_dict['column_names'],
+                    is_unique=index_dict['unique'],
+                )
+            )
+
+        return indexes
+
+    def introspect_primary_key(
+        self, inspector, relation_name, schema_name
+    ) -> Tuple[Optional[str], List[str]]:
+        """Introspect the primary key of a table, returning the pkey name and list of columns in the primary key (if any).
+
+        If no primary key index is defined, will return None for name, empty list for columns.
+        """
+        primary_index_dict = inspector.get_pk_constraint(relation_name, schema_name)
+
+        pkey_name = primary_index_dict.get('name', '(unnamed primary key)')
+
+        if primary_index_dict['constrained_columns']:
+            return pkey_name, primary_index_dict['constrained_columns']
+        else:
+            return '', []  # None, []
+
+    def introspect_columns(
+        self, inspector: 'SchemaStrippingInspector', schema_name: str, relation_name: str
+    ) -> List[ColumnModel]:
+
+        column_dicts = inspector.get_columns(relation_name, schema=schema_name)
+
+        retlist = []
+
+        for col in column_dicts:
+            comment = col.get('comment')  # Some dialects do not return.
+
+            try:
+                type_name = str(col['type'].as_generic()).lower()
+            except NotImplementedError:
+                # ENG-5268: More esoteric types like UUID do not implement .as_generic()
+                type_name = str(col['type']).replace('()', '').lower()
+
+            retlist.append(
+                ColumnModel(
+                    name=col['name'],
+                    is_nullable=col['nullable'],
+                    data_type=type_name,
+                    default_expression=col['default'],
+                    comment=comment,
+                )
+            )
+
+        return retlist
+
+    def inform_gate_start(self, auth_header: dict, datasource_id: UUID):
+        """Tell gate to forget about any prior structures known for this datasource"""
+
+        # No route implemented for this yet, but need one, otherwise Gate-side will
+        # never forget about dropped tables/views.
+
+        pass
+
+    def inform_gate_relation(
+        self,
+        auth_header: dict,
+        datasource_id,
+        relation_description: RelationStructureDescription,
+    ):
+        """POST this `relation_description` up to Gate for storage, relating to `datasource_id`"""
+
+        as_dict_from_json = json.loads(relation_description.json())
+
+        resp = requests.post(
```

Comment:
maybe create a `requests.Session` when creating this class to use for all requests. Since we're accessing the same domain, it'll reuse some connections. You can also set the default headers to avoid needing to set `headers=auth_header` here each time

Reply:
👍